Add filter status to ps module

remyd1 opened the pull request:
### What does this PR do?
It allows to retrieve a list of processes according to their state, using `psutil`.

### What issues does this PR fix or reference?
Fixes: #61420 

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**
- [x] Docs
- [x] Changelog - https://docs.saltproject.io/en/master/topics/development/changelog.html
- [x] Tests written/updated

### Commits signed with GPG?
Yes

Please review [Salt's Contributing Guide](https://docs.saltproject.io/en/master/topics/development/contributing.html) for best practices.

See GitHub's [page on GPG signing](https://help.github.com/articles/signing-commits-using-gpg/) for more information about signing commits with GPG.

remyd1 committed:
add filter status to ps module to get a list of processes according to their state

remyd1 committed:
adding changelog

waynew commented on salt/modules/ps.py:
```diff
@@ -772,3 +772,32 @@ def psaux(name):
+zombie, ...]
+    """
+    ret = []
+    if filter is not None and filter != "":
```
**suggestion** `if filter:` is equivalent to this statement.

remyd1 replied:
Ok, done !

waynew commented on salt/modules/ps.py:
```diff
@@ -772,3 +772,32 @@ def psaux(name):
+    """
+    ret = []
+    if filter is not None and filter != "":
+        sanitize_name = str(filter)
```
**suggestion**  `status` would be a better name
```suggestion
      status = str(filter)
```

It would also need to be changed below.

remyd1 replied:
Done !

waynew reviewed (changes requested):
Hey @remyd1 thanks for the PR! This seems like a useful addition (without checking to see if there's anything else that could accomplish this - I'm assuming that you did that already, given the feature request!)

One thing that this would require is tests.
If you need help writing tests, attend one of the [Test Clinics on Twitch](https://www.twitch.tv/saltprojectoss/schedule)!

remyd1 committed:
modifications from @waynew PR#61421

remyd1 commented:
I would need a bit of help to write the test clinic. I looked at twitch streams but I am still kind of a bit lost.

waynew committed:
add some sample tests

remyd1 committed:
need to add some test coverage to ps module for status filter

remyd1 committed:
reformat

remyd1 commented:
Tests updated !

Ch3LL commented on salt/modules/ps.py:
```diff
@@ -772,3 +773,39 @@ def psaux(name):
+    Returns a list of processes according to their states.
+    See https://psutil.readthedocs.io/en/latest/index.html\
+?highlight=status#process-status-constants
+
```
`.. versionadded:: 3005`

waynew replied:
Well, 3006 now :joy:

waynew replied:
added (and updated the docstring a bit)

Ch3LL reviewed (changes requested):
Just a small change request. Also there are failing tests related to this change that need to be fixed.

Ch3LL commented:
@remyd1 are you able to come back to this PR and fix the tests and fix the feedback submitted?

remyd1 commented:
> @remyd1 are you able to come back to this PR and fix the tests and fix the feedback submitted?

Hi @Ch3LL , unfortunately I have no more time to dig into those test cases. Moreover, I am a bit lost with Mock, MagicMock and associated Salt methods...

If anyone could fix these tests, it would be helpful.

Thanks,
Best regards,

waynew commented:
@remyd1 I can take a look at this during tomorrow's Test Clinic. If you don't have the time/inclination to finish it up I may be able to get it sorted out :+1:

remyd1 commented:
Indeed, if you could check what is wrong here, it would be great 👍🏼

waynew commented:
@remyd1 Didn't I look into this on Tuesday's Test Clinic? Were you able to update the tests here or do you need a bit more help to get this done?

remyd1 commented:
Hi @waynew 

No, sorry. As I said, I have less time to spend on this. Did you do stream some more tests on it ? Could you give me back you Twitch URL ? Anyway, I do not think I will be able to achieve this.

Thanks,

waynew commented:
@remyd1 https://twitch.tv/saltprojectoss

waynew committed:
port modules.ps tests to pytest

waynew committed:
Fix ps.status module tests

Renamed `filter` to `status` to match up with the psutil.Process.status.
Also added a comment to call out the singular status vs multiple, and
mention how that could be added in the future.

Also added versionadded to the docs, re: @Ch3LL's feedback.

Refactored the module a bit to pass tests, as well as simplify the
return bits. There was no need to check that the return type was a
generator - psutil.process_iter is going to yield always, and we don't
really care about anything else getting mocked, we just want to iterate
over it.

Since psutil.Process already has a `.as_dict(attrs=...)` where we can
filter by the attributes we care about, we can go ahead and use that
instead of having to `pop` the ones we don't care about.

waynew commented:
@Ch3LL / @remyd1 I went ahead and updated this on today's Test Clinic. I also took the opportunity to migrate the unittest to pytest since it wasn't super duper involved.

Assuming tests all pass this should be good to go now :+1:

Ch3LL commented:
@waynew looks like there are related test failures

waynew commented:
Doh! Yeah, looks like I didn't understand the API of the Process quite well
enough. Just need to mock a few more things - doing that on today's Test
Clinic!

waynew committed:
Mock more psutil bits

The problem: psutil has Process objects for different platforms, and
those actually try to parse some system output bits that may not exist
when we don't have a process with that PID. Since we're creating one
with PID 42, we need to mock out all the shenanigans that the underlying
Process will use.

Specifically currently that's:

- create_time
- _parse_stat_file
- status

waynew committed:
Go ahead and use extra data

Ch3LL commented:
Looks like just window, mac and windows are failing now

waynew committed:
Go ahead and create _parse_stat_file

On other platforms this attribute doesn't exist, but we don't really
care. If it exists it shouldn't be a problem. So go ahead and
create=True.

waynew committed:
Mock osx and bsd attrs as well

waynew committed:
more bsd time keys

waynew committed:
rename filter to status

Ch3LL commented on salt/modules/ps.py:
```diff
@@ -772,3 +772,56 @@ def psaux(name):
+
+def status(status):
+    """
+    .. versionadded:: 3006
```
```suggestion
    .. versionadded:: 3006.0
```

waynew committed:
Add .0 to version

Co-authored-by: Megan Wilhite <mwilhite@vmware.com>

waynew commented:
@Ch3LL failures didn't *look* related - one looked like the underlying OS/Docker had an issue (completely unrelated to this module) and the others that I saw claimed to be a mem leak issue but... I'm guessing that's also a flaky test :suspect:

Commit:
Merge branch 'master' into add-filter-status-to-ps-module
